man/hfst-compose-intersect.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH HFST-COMPOSE-INTERSECT "1" "April 2026" "HFST" "User Commands"
.SH NAME
hfst-compose-intersect \- =Compose a lexicon with one or more rule transducers.
.SH SYNOPSIS
.B hfst-compose-intersect
[\fI\,OPTIONS\/\fR...] [\fI\,INFILE1 \/\fR[\fI\,INFILE2\/\fR]]
.SH DESCRIPTION
Compose a lexicon with one or more rule transducers.
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.TP
\fB\-\-colour\fR[=\fI\,WHEN\/\fR]
Print in colour WHEN:
.TP
\fB\-\-color\fR[=\fI\,WHEN\/\fR]
always, never, auto (default)
.SS "Input/Output options:"
.TP
\fB\-1\fR, \fB\-\-input1\fR=\fI\,INFILE1\/\fR
Read first input transducer from INFILE1
.TP
\fB\-2\fR, \fB\-\-input2\fR=\fI\,INFILE2\/\fR
Read second input transducer from INFILE2
.TP
\fB\-C\fR, \fB\-\-do\-not\-convert\fR
Do not allow transducers to be converted into the same type
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,OUTFILE\/\fR
Write results to OUTFILE
.SS "Composition options:"
.TP
\fB\-I\fR, \fB\-\-invert\fR
Compose the intersection of the
rules with the lexicon instead
of composing the lexicon with
the intersection of the rules.
.TP
\fB\-f\fR, \fB\-\-fast\fR
Faster compose instersect using
more memory.
.TP
\fB\-e\fR, \fB\-\-encode\-weights\fR
Encode weights when minimizing
(default is false).
.TP
\fB\-a\fR, \fB\-\-harmonize\fR
Harmonize symbols.
.PP
If OUTFILE, or either INFILE1 or INFILE2 is missing or \-, standard
streams will be used. INFILE1, INFILE2, or both, must be specified
The format of INFILE1 and INFILE2 must be the same; the result will
have the same format as these.
INFILE1 (the lexicon) must contain exactly one transducer.
INFILE2 (rule file) may contain several transducers.
.SH EXAMPLES
.IP
hfst\-compose\-intersect \-o analyzer.hfst lexicon.hfst rules.hfst
.PP
compose rules with lexicon
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://github.com/hfst/hfst/issues>
.PP
hfst\-compose\-intersect home page:
<https://github.com/hfst/hfst/wiki/HfstComposeIntersect>
.br
General help using HFST software:
<https://github.com/hfst/hfst/wiki>
.SH COPYRIGHT
Copyright \(co 2017 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# NEWS for hfst-tools and library

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

## Noteworthy changes in 3.17.1

* fixes logic error where `hfst-lexc` would treat a warning about implicit
  alphabet in XREs as a critical error even when `-Werror` is not asked

## Noteworthy changes in 3.17.0

* API breaks: HfstTokenizer, LexcCompiler, `read_in_att_format` functions
  in HfstTransducer, HfstBasicTransducer, etc. to accommodate the following
  unicode handling changes and error reporting:
* Character "tokenisers" that used to operate on bytes or UTF-8 sequences will
  now operate on [graphemes](https://www.unicode.org/reports/tr29/tr29-23.html#Grapheme_Cluster_Boundaries),
  this means that by default arcs in FSAs are expected to match combining
  grapheme clusters, e.g. a base character plus all diacritics.
  The implementation is based on ICU implementation of `u_brk`
    * in lexc string entries, graphemes are compiled into a arcs
    * in xerox regex curly braces, as well
    * in tokenise without --multichar mode, input is tokenised by graphemens
    * etc. etc. tbd
* negative weights yield warnings now unless turned off
* tools will have more gcc-style warning control: -Wall -Werror -Wno-this-error
* lexc warnings are systematically treated as errors with -Werror now
* colourised more warnings and errors

## Noteworthy changes in 3.16.2

* Support for `>OpenFst-1.8.0`
    * only functional change is removal of n-best paths search for log weight
      automata, it probably has errored out before but has slightly different
      error now (HfstException instead of openfst exit)
* Optimisations to hfst-pmatch2fst alphabet handling

## Noteworthy changes in 3.16.1

* Bug fixes:

    * hfst-tokenize escapes both \r and \n now

    * regexp parsing more robust with comments and linebreaks

* Removal of some bundled libraries including foma and openfst, install
  suitable versions to compile support for them

* ICU stabilised as unicode handler over all

* experimental C API for rust bindings (unstable)

* new tools (testing):

    * hfst-check-alpha for checking and comparing alphabets in automata and
      their metadata

    * hfst-dump-alphabets for reviewing alphabets easier than hfst-summarize
      grepping

    * hfst-kill-paths for efficient path removal

* introduced clang-format to work towards standard code formatting

Noteworthy changes in 3.16.0
----------------------------

* Bugfixes:

  * During conversion between optimized-lookup and other formats, the alphabet
    was not being preserved, leading to incorrect behaviour in some cases.
    This has been fixed.

  * Fix segfaults in hfst-summarize, hfst-disjunct and foma back-end.

  * Force ICU, fixes #501.

  * Use unique_ptr for input/ouput stream memory allocation (#547).


* Improvements:

  * hfst-pmatch archives now no longer include or parse redundant alphabets,
    which in some cases can be a meaningful size and speed saving.

  * Multiple memory leaks fixed.

  * Drop support for TR1 standard assuming that C++14 is minimal requirement.

  * Drop legacy wrappers for foma and lexc (#532).

  * Use upstream Foma and by default upstream OpenFST.

  * Use C++17, or C++14 with bundled OpenFST.

  * Update external links for sfst, foma and flex.

  * Add lex/yacc sources to dist (see issue #524).

* Features:

  * A new variable in the hfst-pmatch language, "unicode-character-classes",
    which causes the use of built-in character sets to be interpreted at
    runtime using a unicode library.

  * Verbose output from hfst-pmatch2fst has been much improved, providing eg.
    a call graph of the process. It also warns about use of pmatch features
    that is likely to be incorrect and lead to errors.


Noteworthy changes in 3.15.5
----------------------------

* Fixes to pmatch:

  * Modify tracking of entry points to allow for RTNs in Captures()

  * Fix off-by-one bug in checking captures

* Fixes to other issues: #508, #488, #486, #452, #491, #494

* Fix memory leaks


Noteworthy changes in 3.15.4
----------------------------

* Import Foma patches with license change to Apache-2.0 and minimal other fixes

* Allow regeneration of Foma files so the generated files can be omitted

* Explicit license file for SFST


Noteworthy changes in 3.15.3
----------------------------

* Numerous improvements to command line tools, e.g. substitute, insert-freely, proc, fst2txt, expand


Noteworthy changes in 3.15.2
----------------------------

* Fix off-by-one bug in checking captures in pmatch


Noteworthy changes in 3.15.1
----------------------------

* Fix issue #443

* Fix issue #429

* Fix issue #409


Noteworthy changes in 3.15.0
----------------------------

* Numerous improvements in pmatch processing:

  * hfst-pmatch2fst now implements eg @include"filename.txt" as a preprocessing stage
  
  * hfst-pmatch2fst now implements Explode() and Implode() in full generality
  
  * hfst-pmatch now allows printing weights (in location mode, with --locate)
  
  * hfst.PmatchContainer.locate offered via python bindings
  
  * bugfixes and improvements hfst.PmatchContainer.tokenize

* Add tool hfst-eliminate-flags.

* Allow both escaped and unescaped zeros in multicharacters in all lexc lexicon entries (fixes issue #274).

* Also reweight state that is both initial and final in hfst-reweight (fixes issue #368).

* Do not let distributed files depend on configuration. Include man files in dist.

* Print warnings and error messages also in earlier stages of twol compilation.

* Add option --execute-and-quit (-E) to hfst-xfst.


Noteworthy changes in 3.14.0
----------------------------

* Numerous improvements to pmatching and tokenization:

  * pmatch now supports the default symbol

  * pmatch now supports reading word embeddings in the binary format


tools/src/hfst-compose-intersect.cc
//! @file hfst-compose.cc
//!
//! @brief Transducer composition tool
//!
//! @author HFST Team

//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#ifdef WINDOWS
#include <io.h>
#endif

#include <fstream>
#include <iostream>
#include <memory>

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>
#include <set>

#include "HfstInputStream.h"
#include "HfstOutputStream.h"
#include "HfstTransducer.h"

using hfst::HfstInputStream;
using hfst::HfstOutputStream;
using hfst::HfstTransducer;
using hfst::HfstTransducerVector;
using hfst::ImplementationType;

#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "hfst-tool-metadata.h"

#include "inc/globals-binary.h"
#include "inc/globals-common.h"

// static bool insert_missing_flags=false;

// If invert is true, the intersection of the rules is composed eith
// the lexicon. Otherwise the lexicon is composed with the
// intersection of the rules.
static bool invert = false;
static bool encode_weights = false;
static bool fast_ci = false;
static bool harmonize = false;

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out,
            "Usage: %s [OPTIONS...] [INFILE1 [INFILE2]]\n"
            "Compose a lexicon with one or more rule transducers.\n"
            "\n",
            program_name);
    print_common_program_options(message_out);
    print_common_binary_program_options(message_out);
    fprintf(message_out,
            "Composition options:\n"
            "  -I, --invert                 Compose the intersection of the\n"
            "                               rules with the lexicon instead\n"
            "                               of composing the lexicon with\n"
            "                               the intersection of the rules.\n"
            "  -f, --fast                   Faster compose instersect using\n"
            "                               more memory.\n"
            "  -e, --encode-weights         Encode weights when minimizing\n"
            "                               (default is false).\n"
            "  -a, --harmonize              Harmonize symbols.\n");
    // print_common_binary_program_parameter_instructions(message_out);
    fprintf(
        message_out,
        "\nIf OUTFILE, or either INFILE1 or INFILE2 is missing or -, "
        "standard\n"
        "streams will be used. INFILE1, INFILE2, or both, must be specified\n"
        "The format of INFILE1 and INFILE2 must be the same; the result will\n"
        "have the same format as these.\n"
        "INFILE1 (the lexicon) must contain exactly one transducer.\n"
        "INFILE2 (rule file) may contain several transducers.\n");
    fprintf(message_out,
            "\n"
            "Examples:\n"
            "  %s -o analyzer.hfst lexicon.hfst rules.hfst\n"
            "compose rules with lexicon\n"
            "\n",
            program_name);
    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
}

int
parse_options(int argc, char **argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[]
            = { HFST_GETOPT_COMMON_LONG,
                HFST_GETOPT_BINARY_LONG,
                { "invert", no_argument, 0, 'I' },
                { "encode-weights", no_argument, 0, 'e' },
                { "fast", no_argument, 0, 'f' },
                { "harmonize", no_argument, 0, 'a' },
                { 0, 0, 0, 0 } };
        int option_index = 0;
        int c = getopt_long(argc, argv,
                            HFST_GETOPT_COMMON_SHORT HFST_GETOPT_BINARY_SHORT
                            "FIeHfa",
                            long_options, &option_index);
        if (-1 == c)
        {
            break;
        }
        switch (c)
        {
#include "inc/getopt-cases-binary.h"
#include "inc/getopt-cases-common.h"
#include "inc/getopt-cases-error.h"
        case 'I':
            invert = true;
            break;
        case 'e':
            encode_weights = true;
            break;
        case 'f':
            fast_ci = true;
            break;
        case 'a':
            harmonize = true;
            break;
        }
    }

#include "inc/check-params-binary.h"
#include "inc/check-params-common.h"
    return EXIT_CONTINUE;
}

using hfst::HfstTokenizer;
using hfst::implementations::HfstBasicTransducer;
using hfst::implementations::HfstState;

typedef std::set<std::string> StringSet;

bool
is_special_symbol(const std::string &symbol)
{
    return symbol.size() > 2 and symbol[0] == '@'
           and *(symbol.rbegin()) == '@';
}

std::string
check_all_symbols(const HfstTransducer &lexicon, const HfstTransducer &rule)
{
    HfstBasicTransducer rule_b(rule);

    StringSet rule_input_symbols;

    for (HfstState s = 0; s <= rule_b.get_max_state(); ++s)
    {
        for (hfst::implementations::HfstBasicTransitions::const_iterator it
             = rule_b[s].begin();
             it != rule_b[s].end(); ++it)
        {
            const std::string &input_symbol = it->get_input_symbol();
            rule_input_symbols.insert(input_symbol);
        }
    }

    if (rule_input_symbols.count(hfst::internal_identity) != 0)
    {
        return "";
    }

    HfstBasicTransducer lexicon_b(lexicon);

    for (HfstState s = 0; s <= lexicon_b.get_max_state(); ++s)
    {
        for (hfst::implementations::HfstBasicTransitions::const_iterator it
             = lexicon_b[s].begin();
             it != lexicon_b[s].end(); ++it)
        {
            const std::string &output_symbol = it->get_output_symbol();

            if (rule_input_symbols.count(output_symbol) == 0)
            {
                return output_symbol;
            }
        }
    }

    return "";
}

std::string
check_multi_char_symbols(const HfstTransducer &lexicon,
                         const HfstTransducer &rule)
{
    HfstBasicTransducer lexicon_b(lexicon);
    HfstBasicTransducer rule_b(rule);

    HfstTokenizer tokenizer;

    StringSet rule_input_symbols;

    for (HfstState s = 0; s <= rule_b.get_max_state(); ++s)
    {
        for (hfst::implementations::HfstBasicTransitions::const_iterator it
             = rule_b[s].begin();
             it != rule_b[s].end(); ++it)
        {
            const std::string &input_symbol = it->get_input_symbol();
            rule_input_symbols.insert(input_symbol);
        }
    }

    for (HfstState s = 0; s <= lexicon_b.get_max_state(); ++s)
    {
        for (hfst::implementations::HfstBasicTransitions::const_iterator it
             = lexicon_b[s].begin();
             it != lexicon_b[s].end(); ++it)
        {
            const std::string &output_symbol = it->get_output_symbol();

            if (rule_input_symbols.count(output_symbol) == 0)
            {
                if (is_special_symbol(output_symbol))
                {
                    continue;
                }

                if (tokenizer.tokenize_one_level(output_symbol, false).size()
                    > 1)
                {
                    return output_symbol;
                }
            }
        }
    }

    return "";
}

void
harmonize_rules(HfstTransducer &lexicon, std::vector<HfstTransducer> &rules)
{
    for (std::vector<HfstTransducer>::iterator it = rules.begin();
         it != rules.end(); it++)
    {
        it->harmonize(lexicon);
    }
    return;
}

int
compose_streams(HfstInputStream &firststream, HfstInputStream &secondstream)
{
    // there must be at least one transducer in both input streams
    // bool continueReading = firststream.is_good() && secondstream.is_good();

    hfst::ImplementationType type1 = firststream.get_type();
    hfst::ImplementationType type2 = secondstream.get_type();
    hfst::ImplementationType output_type = hfst::UNSPECIFIED_TYPE;
    if (type1 != type2)
    {
        if (allow_transducer_conversion)
        {
            int ct = conversion_type(type1, type2);
            std::string warnstr("Transducer type mismatch in "
                                + std::string(firstfilename) + " and "
                                + std::string(secondfilename) + "; ");
            if (ct == 1)
            {
                warnstr.append("using former type as output");
                output_type = type1;
            }
            else if (ct == 2)
            {
                warnstr.append("using latter type as output");
                output_type = type2;
            }
            else if (ct == -1)
            {
                warnstr.append("using former type as output, loss of "
                               "information is possible");
                output_type = type1;
            }
            else /* should not happen */
            {
                throw "Error: hfst-compose-intersect: conversion_type "
                      "returned an invalid integer";
            }
            hfst_warning(0, 0, warnstr.c_str());
        }
        else
        {
            hfst_error(
                EXIT_FAILURE, 0,
                "Transducer type mismatch in %s and %s; "
                "formats %s and %s are not compatible for compose-intersect "
                "(--do-not-convert was requested)",
                firstfilename, secondfilename, hfst_strformat(type1),
                hfst_strformat(type2));
        }
    }
    else
    {
        output_type = type1;
    }

    HfstOutputStream outstream
        = (outfile != stdout) ? HfstOutputStream(outfilename, output_type)
                              : HfstOutputStream(output_type);

    bool bothInputs = firststream.is_good() && secondstream.is_good();
    (void)bothInputs;

    if (is_input_stream_in_ol_format(firststream, "hfst-compose-intersect")
        || is_input_stream_in_ol_format(secondstream,
                                        "hfst-compose-intersect"))
    {
        return EXIT_FAILURE;
    }

    HfstTransducerVector rules;
    size_t rule_n = 1;

    while (secondstream.is_good())
    {
        HfstTransducer rule(secondstream);
        rule.convert(output_type);
        const char *rulename = strdup(rule.get_name().c_str());
        if (strlen(rulename) > 0)
        {
            verbose_printf("Reading and minimizing rule %s...\n", rulename);
        }
        else
        {
            verbose_printf("Reading and minimizing rule " SIZE_T_SPECIFIER
                           "...\n",
                           rule_n);
        }
        bool enc = hfst::get_encode_weights();
        if (encode_weights)
        {
            hfst::set_encode_weights(true);
        }
        rule.minimize();
        if (encode_weights)
        {
            hfst::set_encode_weights(enc);
        }

        rules.push_back(rule);
        rule_n++;
    }

    while (firststream.is_good())
    {
        verbose_printf("Reading lexicon...");
        HfstTransducer lexicon(firststream);
        lexicon.convert(output_type);
        char *lexiconname = hfst_get_name(lexicon, firstfilename);
        verbose_printf(" %s read\n", lexiconname);

        verbose_printf("Computing intersecting composition...\n");

        if (rules.size() > 0)
        {
            std::string symbol;
            if ((symbol = check_all_symbols(lexicon, rules[0])) != "")
            {
                hfst_warning(
                    0, 0,
                    "\nFound output symbols (e.g. \"%s\") in transducer in\n"
                    "file %s which will be filtered out because they are\n"
                    "not found on the input tapes of transducers in file\n"
                    "%s.",
                    symbol.c_str(), firstfilename, secondfilename);
            }
            else if ((symbol = check_multi_char_symbols(lexicon, rules[0]))
                     != "")
            {
                hfst_warning(
                    0, 0,
                    "\nFound output multi-char symbols (\"%s\") in \n"
                    "transducer in file %s which are not found on the\n"
                    "input tapes of transducers in file %s.",
                    symbol.c_str(), firstfilename, secondfilename);
            }
        }

        if (harmonize)
        {
            harmonize_rules(lexicon, rules);
        }

        if (fast_ci)
        {
            // To hopefully speed up stuff: Compose intersect the output
            // of the lexicon with the rules and then compose the original
            // lexicon with the result.

            if (invert)
            {
                HfstTransducer lexicon_input(lexicon);
                lexicon_input.input_project().minimize();
                lexicon_input.compose_intersect(rules, true);

                lexicon_input.compose(lexicon);
                lexicon = lexicon_input;
            }
            else
            {
                HfstTransducer lexicon_output(lexicon);
                lexicon_output.output_project().minimize();
                lexicon_output.compose_intersect(rules, false);
                lexicon.compose(lexicon_output);
            }
        }
        else
        {
            lexicon.compose_intersect(rules, invert);
        }

        char *composed_name = static_cast<char *>(
            malloc(sizeof(char)
                       * (strlen(lexiconname) + strlen(secondfilename)
                          + strlen("compose(%s, interserct(%s))"))
                   + 1));
        if (sprintf(composed_name, "compose(%s, intersect(%s))", lexiconname,
                    secondfilename)
            > 0)
        {
            lexicon.set_name(composed_name);
        }
        hfst_set_formula(lexicon, lexicon, " ∘ ⋂R");

        verbose_printf("Storing result in %s...\n", outfilename);
        outstream << lexicon;
    }

    firststream.close();
    secondstream.close();
    outstream.close();
    return EXIT_SUCCESS;
}

int
main(int argc, char **argv)
{
#ifdef WINDOWS
    _setmode(0, _O_BINARY);
    _setmode(1, _O_BINARY);
#endif
    hfst_set_program_name(argv[0], "0.1", "HfstComposeIntersect");
    int retval = parse_options(argc, argv);
    if (retval != EXIT_CONTINUE)
    {
        return retval;
    }
    // close buffers, we use streams
    if (firstfile != stdin)
    {
        fclose(firstfile);
    }
    if (secondfile != stdin)
    {
        fclose(secondfile);
    }
    if (outfile != stdout)
    {
        fclose(outfile);
    }
    verbose_printf("Reading from %s and %s, writing to %s\n", firstfilename,
                   secondfilename, outfilename);
    // here starts the buffer handling part
    std::unique_ptr<HfstInputStream> firststream;
    std::unique_ptr<HfstInputStream> secondstream;
    try
    {
        firststream.reset(firstfile != stdin
                              ? new HfstInputStream(firstfilename)
                              : new HfstInputStream());
    }
    catch (const HfstException e)
    {
        hfst_error(EXIT_FAILURE, 0, "%s is not a valid transducer file",
                   firstfilename);
    }
    try
    {
        secondstream.reset((secondfile != stdin)
                               ? new HfstInputStream(secondfilename)
                               : new HfstInputStream());
    }
    catch (const HfstException e)
    {
        hfst_error(EXIT_FAILURE, 0, "%s is not a valid transducer file",
                   secondfilename);
    }

    retval = compose_streams(*firststream, *secondstream);
    free(firstfilename);
    free(secondfilename);
    free(outfilename);
    return retval;
}
